### PR TITLE
feat: add reduceIndexed

### LINKED ISSUE #281: reduceIndexed
```js
// indexedReduce :: ((a, b, c, [b]) -> a) -> a -> [b] -> a
export const indexedReduce = addIndex(reduce);

### PR DESCRIPTION
**Pull request by guillaumearm**
Closes #281

### CODE REVIEW

File: src/reduceIndexed.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * `reduceIndexed` implementation is simple : `
+ * const reduceIndexed = R.addIndex(R.reduce);
+ * `
+ * @func reduceIndexed
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/2.5.0|v2.5.0}
+ * @category List
+ * @typedef Idx = Number
+ * @sig ((a, b, Idx, [b]) => a) -> a -> [b] -> a
+ * @param {Function} fn The function to be called on every element of the input `list`
```
Comment (char0n) [inline src/reduceIndexed.js]: The iterator function. Receives four values, the accumulator, the current element from the array, index and the entire list

File: src/reduceIndexed.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * @param {*} acc The accumulator value
+ * @param {Array} list The list iterate over
```
Comment (char0n) [inline src/reduceIndexed.js]: The list to iterate over

File: src/reduceIndexed.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * @return {*} The final, accumulated value
+ * @see {@link http://ramdajs.com/docs/#addIndex|R.addIndex}, {@link http://ramdajs.com/docs/#reduce|R.reduce}
+ * @example const initialList = ['f', 'o', 'o', 'b', 'a', 'r'];
```
Comment (char0n) [inline src/reduceIndexed.js]: Format this correctly as I've done here: https://github.com/char0n/ramda-adjunct/commit/318cf023d3aa2dd71cfd9bfc2ebc8eec48220058

**Comment by char0n** [comment]
@guillaumearm the tests are not passing. We are supporting ramda from 0.19.0 to 0.25.0. I guess there are some incompatibilities there between various versions.

To correct this you'll have to feature detection as we've for example here: https://github.com/char0n/ramda-adjunct/pull/260#issuecomment-355324315

**Comment by guillaumearm** [comment]
@char0n,

The broken test : `Prefers the use of the iterator of an object over reduce (and handles short-circuits)` is about testing this : https://github.com/ramda/ramda/commit/103543e2e1c2b8a43275fff99c9e62363d1def85

In this case, the feature detection IS the test itself, I think we can just replace this test by a simple reduce test on iterable.

File: src/index.d.ts [hunk 4]
```diff
@@ -525,6 +525,23 @@ declare namespace RamdaAdjunct {
...
+            acc: TResult,
+            list: R,
+        ): TResult;
+        reduceIndexed<T, TResult, R extends T[]>(
+            iterator: (acc: TResult, elem: T, key: number, list: R) => TResult,
+            acc: TResult,
+        ): (list: R) => TResult;
+        reduceIndexed<T, TResult, R extends T[]>(
+            iterator: (acc: TResult, elem: T, key: number, list: R) => TResult,
+        ): (acc: TResult) => (list: R) => TResult;
```
Comment (BjornMelgaard) [inline src/index.d.ts]: https://www.typescriptlang.org/play/index.html#src=declare%20function%20reduceIndexed%3CT%2C%20TResult%2C%20R%20extends%20T%5B%5D%3E(%0D%0A%20%20%20%20iterator%3A%20(acc%3A%20TResult%2C%20elem%3A%20T%2C%20key%3A%20number%2C%20list%3A%20R)%20%3D%3E%20TResult%2C%0D%0A%20%20%20%20acc%3A%20TResult%2C%0D%0A%20%20%20%20list%3A%20R%2C%0D%0A)%3A%20TResult%3B%0D%0Adeclare%20function%20reduceIndexed%3CT%2C%20TResult%2C%20R%20extends%20T%5B%5D%3E(%0D%0A%20%20%20%20iterator%3A%20(acc%3A%20TResult%2C%20elem%3A%20T%2C%20key%3A%20number%2C%20list%3A%20R)%20%3D%3E%20TResult%2C%0D%0A%20%20%20%20acc%3A%20TResult%2C%0D%0A)%3A%20(list%3A%20R)%20%3D%3E%20TResult%3B%0D%0Adeclare%20function%20reduceIndexed%3CT%2C%20TResult%2C%20R%20extends%20T%5B%5D%3E(%0D%0A%20%20%20%20iterator%3A%20(acc%3A%20TResult%2C%20elem%3A%20T%2C%20key%3A%20number%2C%20list%3A%20R)%20%3D%3E%20TResult%2C%0D%0A)%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20(acc%3A%20TResult)%3A%20(list%3A%20R)%20%3D%3E%20TResult%3B%0D%0A%20%20%20%20%20%20%20%20(acc%3A%20TResult%2C%20list%3A%20R)%3A%20TResult%0D%0A%7D%3B%0D%0A%0D%0Alet%20x1%20%3D%20reduceIndexed((acc%3A%20number%2C%20elem%3A%20number)%20%3D%3E%20acc%20%2B%20elem%2C%200%2C%20%5B1%2C%202%2C%203%5D)%0D%0Alet%20x2%20%3D%20reduceIndexed((acc%3A%20number%2C%20elem%3A%20number)%20%3D%3E%20acc%20%2B%20elem%2C%200)(%5B1%2C%202%2C%203%5D)%0D%0Alet%20x3%20%3D%20reduceIndexed((acc%3A%20number%2C%20elem%3A%20number)%20%3D%3E%20acc%20%2B%20elem)(0%2C%20%5B1%2C%202%2C%203%5D)%0D%0Alet%20x4%20%3D%20reduceIndexed((acc%3A%20number%2C%20elem%3A%20number)%20%3D%3E%20acc%20%2B%20elem)(0)(%5B1%2C2%2C3%5D)%0D%0A

you should add one case

Good work

File: test/reduceIndexed.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+      eq(R.reduce(rfn, [], list), [1, 2]);
+    });
+  });
+
+  context('documentation example', function () {
+    specify('should join idx and value with a "-"', function () {
+      const initialList = ['f', 'o', 'o', 'b', 'a', 'r'];
+      const resultAcc = '-f0-o1-o2-b3-a4-r5';
+      eq(RA.reduceIndexed((acc, val, idx) => `${acc}-${val}${idx}`, '', initialList), resultAcc);
```
Comment (char0n) [inline test/reduceIndexed.js]:
```js
const initialList = ['f', 'o', 'o', 'b', 'a', 'r'];
const resultAcc = '-f0-o1-o2-b3-a4-r5';

eq(RA.reduceIndexed((acc, val, idx) => `${acc}-${val}${idx}`, '', initialList), resultAcc);
```
Insert empty line here fo better readability

File: test/reduceIndexed.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    });
+  });
+
+  context('fourth argument', function () {
+    const initialList = ['f', 'o', 'o', 'b', 'a', 'r'];
+    specify('should be the same reference that original list', function () {
```
Comment (char0n) [inline test/reduceIndexed.js]: Insert empty line between const and specify to increase readability

**Comment by char0n** [comment]
Fix the last two comments and we can merge. Nice work !

**Comment by guillaumearm** [comment]
Yup thanks, It's done.

All of this inspire me to add more specific lint rules to avoid wasting time with these kinds of things.

**Comment by char0n** [comment]
Well not sure if something like that would be possible. It depends on the context and also of the subjective opinion of the reviewer.

**Pull request merged**